### ISSUE TITLE
[Tracking][Performance][AMD] Qwen3 & Qwen3-VL Latency Optimization on AMD MI300 Series GPUs

### ISSUE THREAD

**Issue by wuhuikx** [sgl-project/sglang#18466]
## Description

We aim to optimize the inference latency of **Qwen3-235B** and **Qwen3-VL-235B** models on AMD MI300 series GPU platforms using the [SGLang](https://github.com/sgl-project/sglang) inference engine. This issue serves as a centralized hub to track all relevant PRs, facilitating a comprehensive understanding of the optimization progress.

### Target Models
- **Qwen3-235B** (MoE architecture, 235B total params, 22B active params)
- **Qwen3-VL-235B** (Vision-Language model, same LLM backbone + ViT encoder)

### Hardware Platform
- AMD MI300 series GPUs (CDNA 3 architecture)
- Single node, 8x GPUs with TP8 deployment

### Quantization
- PTPC FP8 (Per-Token-Activation, Per-Channel-Weight) quantization
- Model weights: [Qwen3-235B-FP8](https://huggingface.co/RedHatAI/Qwen3-235B-A22B-FP8-dynamic), [Qwen3-VL-235B-FP8](https://huggingface.co/RedHatAI/Qwen3-VL-235B-A22B-Instruct-FP8-dynamic)

---

## Optimization Overview

---

## 1. Profiling & Debug Tools

Profiling tools are essential for identifying performance bottlenecks. We developed host process profiling capabilities to gain visibility into CPU-side operations that were previously invisible.

| Optimization           | Description                                                  | PR     |
| ---------------------- | ------------------------------------------------------------ | ------ |
| Host Process Profiling | Add profiling support for TokenizerManager to capture CPU-side operations (tokenization, image preprocessing, request handling) alongside GPU operations. Enables end-to-end latency analysis for multimodal workloads. | #16809 |

## 2. LLM Compute Kernel Optimization

These optimizations target the core computation modules of the Qwen3-235B LLM backbone, leveraging AMD AITER library kernels and operator fusion to reduce HBM access and kernel launch overhead.

| Optimization                                    | Description                                                  | PR     |
| ----------------------------------------------- | ------------------------------------------------------------ | ------ |
| HIPBlas Swizzle BF16 & FP8 GEMM                 | Integrate HIPBlas swizzle BF16 and FP8 GEMM kernels for optimized matrix multiplication on AMD GPUs. Applied to both LLM linear layers and ViT computation. | #17161 |
| RMSNorm + FP8 Dynamic Quant Fusion              | Fuse RMSNorm and FP8 dynamic quantization kernels to reduce memory access overhead. Achieves 1-6% E2E latency improvement and 1-2% throughput gain. | #16485 |
| AllReduce + Residual RMSNorm + FP8 Quant Fusion | Fuse AllReduce, residual addition, RMSNorm, and FP8 dynamic quantization into a single kernel, reducing HBM access frequency. | #17262 |
| Fused QKNorm + RoPE + KV Set + Quant + Shuffle  | Integrate AITER's fused QKNorm + RoPE + KV buffer set kernel for Qwen3MoE/Qwen3-VL attention layers, providing significant per-layer performance improvements. | #17223 |

---

## 3. VLM/VIT Optimization

Qwen3-VL-235B introduces additional latency from multimodal data preprocessing, data transport, and ViT encoder computation. These optimizations target all three aspects.

### 3.1 Multimodal Preprocessing Acceleration

| Optimization                                   | Description                                                  | PR                             |
| ---------------------------------------------- | ------------------------------------------------------------ | ------------------------------ |
| Image Decode Acceleration (PIL -> torchvision) | Replace serial PIL-based image decoding in Transformers with parallel torchvision decoding in SGLang's multi-threaded pool. | #17208                         |
| rocJPEG GPU-Accelerated JPEG Decode            | Integrate [rocJPEG](https://github.com/ROCm/rocJPEG) as torchvision backend for GPU-accelerated JPEG decoding on AMD GPUs. Single 720P image decode: ~27ms -> ~4ms (**~7x speedup**). | #17218, pytorch/vision#9342    |
| Remove Redundant Text Tokenization             | Eliminate unnecessarily repeated text tokenization in multimodal model scenarios, reducing host-side preprocessing overhead. | #17206                         |
| Reduce Redundant \<pad\> Tokenization          | Tokenize only a single \<image_pad\>/\<video_pad\> and insert them as n*grid_thw, accelerating Qwen3VL tokenization process. *(Upstream to HuggingFace Transformers)* | huggingface/transformers#43297 |
| Multimodal Processor Parameters                | Add `--mm-processor-kwargs` parameter (similar to vLLM) to manually specify multimodal processor parameters like `min_pixels`/`max_pixels`, enabling convenient tuning and acceleration. | #17131                         |
| Eliminate Unnecessary CPU/D2D Ops in ViT       | Remove unnecessary `contiguous()` calls in VisionAttention and eliminate D2D operations in `torch.where` and `index_put` during ViT forward pass. | #17202                         |
| Accelerate Multimodal Hash Encoding            | Eliminate concat CPU operations in hash encoding, use faster hash algorithm (consistent with vLLM), and offload hash computation to GPU on AMD platforms. | #17203                         |

### 3.2 VisionAttention Optimization

| Optimization                       | Description                                                  | PR     |
| ---------------------------------- | ------------------------------------------------------------ | ------ |
| VisionAttention H2D Precomputation | Precompute `cu_seqlens` and `max_seqlen` metadata once at the beginning of ViT forward pass instead of redundantly in every vision block, introducing LLM-Attention-like metadata into VisionAiterAttention. | #16705 |

### 3.3 VIT Compute Kernel Optimization

| Optimization                            | Description                                                  | PR     |
| --------------------------------------- | ------------------------------------------------------------ | ------ |
| Add & LayerNorm Fusion for Qwen3-VL ViT | Fuse Add and LayerNorm operations in the ViT encoder of Qwen3-VL model, reducing kernel launch overhead and HBM access. | #16571 |

---

## 4. Attention Optimization

Attention module optimizations focus on KV-Cache layout and data type optimizations for the AMD CDNA 3 architecture.

| Optimization                                                 | Description                                                  | PR          |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ----------- |
| Integrate AITER `pa_persistent_fwd` & unified KV-Cache layout | Adopt a unified preshuffle 5D KV-Cache layout (`k_cache: [num_blocks, num_kv_heads, head_dim//x, block_size, x]`, `v_cache: [num_blocks, num_kv_heads, block_size//x, head_dim, x]`) shared by both prefill and decode, aligned with AMD CDNA memory access patterns. Replace `paged_attention_ragged` with AITER's `pa_persistent_fwd` for decode, and use `mha_batch_prefill_func` with the same paged KV-Cache for prefill, Includes native FP8 KV-Cache support. | Coming soon |

---

## 5. Memory Management

Memory management optimizations focus on efficient multimodal data transport between processes.

| Optimization                       | Description                                                  | PR                     |
| ---------------------------------- | ------------------------------------------------------------ | ---------------------- |
| CUDA IPC Multimodal Data Transport | ROCm backend supports the existing CUDA IPC transport feature, enabling GPU-to-GPU direct data transfer between Tokenizer and Scheduler processes without CPU-mediated copies. For 5 images at 960x1280, reduces transport latency by up to **2 seconds**. | N/A (upstream feature) |
| CUDA Kernel Allocator              | Implement memory allocation using CUDA kernels to reduce Python-side overhead in memory management improving allocation efficiency . | Coming soon            |

## 6. Distributed Parallelism

Distributed parallelism optimizations focus on ViT encoder data parallelism to alleviate the compute bottleneck caused by high-resolution images and multi-image inputs.

| Optimization                                               | Description                                                  | PR     |
| ---------------------------------------------------------- | ------------------------------------------------------------ | ------ |
| Qwen3-VL Vision Model DP Support                           | Add data parallel support for Qwen3-VL vision encoder, distributing multi-image ViT computation across GPUs. 3-6% TTFT improvement with 4 images. | #13724 |
| ViT DP Enhancement: Inter-request DP & Single-image Bypass | Enable ViT DP across requests and process single-image requests directly without DP overhead. Includes VisionAttention H2D precomputation. | #17192 |

## 7. Bug Fixes

| Bug Fix                                    | Description                                                  | PR     |
| ------------------------------------------ | ------------------------------------------------------------ | ------ |
| bench_serving Multimodal Image Error       | Fix bench_serving.py error when randomly generated tokens produce image tokens, causing multiple-image assumption mismatches. Affects both NVIDIA and AMD GPUs. | #13254 |
| TTFT Measurement Error with Special Tokens | Fix inaccurate TTFT measurement when special tokens like `<video_pad>` are randomly generated in bench_serving.py. Ensures no special tokens are randomly generated, preventing empty response returns. | #16645 |

---

### Server Launch
```bash
export SGLANG_USE_AITER=1
model=/path/to/Qwen3-VL-235B-A22B-Instruct-FP8-dynamic/
TP=8

python3 -m sglang.launch_server \
    --model-path ${model} \
    --host localhost \
    --port 9000 \
    --tp-size ${TP} \
    --trust-remote-code \
    --chunked-prefill-size 16384 \
    --mem-fraction-static 0.8 \
    --disable-radix-cache \
    --max-prefill-tokens 16384 \
    --cuda-graph-max-bs 128 \
    --max-running-requests 128 \
    --mm-attention-backend aiter_attn \
    --mm-enable-dp-encoder \
    --enable-aiter-allreduce-fusion
```

### Latency Benchmark
```bash
python3 -m sglang.bench_serving \
    --backend sglang-oai-chat \
    --port 9000 \
    --dataset-name image \
    --num-prompts 3 \
    --image-count 5 \
    --image-resolution 960x1280 \
    --random-input-len 8000 \
    --random-output-len 5 \
    --max-concurrency 1 \
    --random-range-ratio 1.0
```

### Accuracy Benchmark (MMMU)
```bash
python3 benchmark/mmmu/bench_sglang.py --port 9000 --concurrency 16
```

---

## References

- [AITER Library (AMD)](https://github.com/ROCm/aiter)
- [SGLang Documentation](https://docs.sglang.io/)
- [Qwen3-235B Model](https://huggingface.co/Qwen/Qwen3-235B-A22B)
- [Qwen3-VL-235B Model](https://huggingface.co/Qwen/Qwen3-VL-235B-A22B-Instruct)
- [rocJPEG (AMD)](https://github.com/ROCm/rocJPEG)
